### PR TITLE
Add WalletConnect support

### LINKED ISSUE #1959: wallet-connect based authentication
to get wallet-connect working for authentication (for now, but we also hope it works for transaction signing which needs a bit of design).

You can hide this feature in a menu and we will first test it and then come up with the right ux design. 

experiment by trying to login through QR code on https://localcryptos.com/login

Familiar yourself with wallet-connect. You can get a wallet connect QR code to start with, by attempting to login to websites that uses it, like https://localcryptos.com/login

You will get a QR code like this:

Screenshot_2020-06-10 Log In To LocalCryptos

Which, if decrypted, has this link:

    wc:a97c80fa-e54a-4346-a805-25c293ac1c4c@1?bridge=https://bridge.walletconnect.org&key=14c4225e2cafddf44e0508fc2c2769853f396734c1b6efd3802f62bc51b2b667

### PR DESCRIPTION
**Pull request by vladyslav-iosdev**
Closes #1959

### CODE REVIEW

**Comment by vladyslav-iosdev** [comment]
For now it allows app act like server only.

File: AlphaWallet/WalletConnect/ViewController/SessionListViewController.swift
Comment (hboon) [reply AlphaWallet/WalletConnect/ViewController/SessionListViewController.swift:10]: Don't we all miss proper Swift namespacing? :P

File: Podfile
Comment (hboon) [inline Podfile:1]: Is this needed to get the pods to work? Otherwise let's not change it in the PR.

File: AlphaWallet/WalletConnect/ViewController/SessionDetailsViewController.swift
Comment (hboon) [inline AlphaWallet/WalletConnect/ViewController/SessionDetailsViewController.swift:161]: "disconnect". A few occurrences.

File: AlphaWallet/WalletConnect/WalletConnectServer.swift
Comment (hboon) [inline AlphaWallet/WalletConnect/WalletConnectServer.swift:36]: Does this need to be dynamic?

**Comment by vladyslav-iosdev** [comment]
> After I scan a WalletConnect QR code, tap start in the app and the session ID appears in the new "WalletConnect" screen, the browser says "Request rejected". Do you know why?
> 
> Otherwise, yes, seems like the right direction. Thanks.

Yep, it rejects all requests for now.

File: Podfile
Comment (vladyslav-iosdev) [reply Podfile:1]: i changed because it doesnt work without this update.

File: AlphaWallet/WalletConnect/WalletConnectServer.swift
Comment (vladyslav-iosdev) [reply AlphaWallet/WalletConnect/WalletConnectServer.swift:36]: now it doesn't, i was need to handle `sessions` updates, just missed to remove it.

File: AlphaWallet/WalletConnect/ViewController/SessionListViewController.swift
Comment (vladyslav-iosdev) [reply AlphaWallet/WalletConnect/ViewController/SessionListViewController.swift:10]: @hboon do i need to update request handlers (for example if we receive `eth_signTransaction` request or eth.) ?

Comment (hboon) [reply AlphaWallet/WalletConnect/ViewController/SessionListViewController.swift:10]: You can check the file “WKWebConfiguration” (I think) for how it’s setting handlers for browsers. I imagine you might have to have a duplicate of that (but the browser version might be a bit more complicated). Maybe some of those can be refactored and extracted. But let’s do the refactoring a bit later.

**Comment by hboon** [comment]
@vladyslav-iosdev are you waiting for me to review this?

**Comment by vladyslav-iosdev** [comment]
> @vladyslav-iosdev are you waiting for me to review this?

this one not finished yet, but u can review it.

**Comment by hboon** [comment]
@vladyslav-iosdev I might come back to it later. Don't wait for me :)

**Comment by vladyslav-iosdev** [comment]
@hboon can u share example signTypedData json  that we receive in dapp action? becauser it seems like we have different json in example wallet connect app.
here is what we got in example:

```
const example = {
  types: {
    EIP712Domain: [
      { name: "name", type: "string" },
      { name: "version", type: "string" },
      { name: "verifyingContract", type: "address" },
    ],
    RelayRequest: [
      { name: "target", type: "address" },
      { name: "encodedFunction", type: "bytes" },
      { name: "gasData", type: "GasData" },
      { name: "relayData", type: "RelayData" },
    ],
    GasData: [
      { name: "gasLimit", type: "uint256" },
      { name: "gasPrice", type: "uint256" },
      { name: "pctRelayFee", type: "uint256" },
      { name: "baseRelayFee", type: "uint256" },
    ],
    RelayData: [
      { name: "senderAddress", type: "address" },
      { name: "senderNonce", type: "uint256" },
      { name: "relayWorker", type: "address" },
      { name: "paymaster", type: "address" },
    ],
  },
  domain: {
    name: "GSN Relayed Transaction",
    version: "1",
    chainId: 42,
    verifyingContract: "0x6453D37248Ab2C16eBd1A8f782a2CBC65860E60B",
  },
  primaryType: "RelayRequest",
  message: {
    target: "0x9cf40ef3d1622efe270fe6fe720585b4be4eeeff",
    encodedFunction:
      "0xa9059cbb0000000000000000000000002e0d94754b348d208d64d52d78bcd443afa9fa520000000000000000000000000000000000000000000000000000000000000007",
    gasData: { gasLimit: "39507", gasPrice: "1700000000", pctRelayFee: "70", baseRelayFee: "0" },
    relayData: {
      senderAddress: "0x22d491bde2303f2f43325b2108d26f1eaba1e32b",
      senderNonce: "3",
      relayWorker: "0x3baee457ad824c94bd3953183d725847d023a2cf",
      paymaster: "0x957F270d45e9Ceca5c5af2b49f1b5dC1Abb0421c",
    },
  },
};
```

**Comment by hboon** [comment]
Hmm. Skip EIP712 for now. It might be more complicated.

**Comment by vladyslav-iosdev** [comment]
@hboon can u review this pr?

**Comment by vladyslav-iosdev** [comment]
@hboon removed unnecessary changes in project file.

**Comment by hboon** [comment]
@James-Sangalli do you have bandwidth to follow up on this?

**Comment by vladyslav-iosdev** [comment]
@hboon i fixed parsing actions according to WalletConnect documentation, checked with example app they provided. 
https://docs.walletconnect.org/json-rpc-api-methods/ethereum#example-parameters-1

**Comment by James-Sangalli** [comment]
@vladyslav-iosdev when I scan this example QR code: https://miro.medium.com/max/2880/0*KnY8fy140-wNRXE2, the wallet does nothing but the output in the console says 

`WC: didConnect url: https://platform.neufund.org/api/wc-bridge-socket/
WC: ==> {"payload":"","topic":"5b16b995-0cef-4084-a966-b0c0677022d6","type":"sub"}`

File: AlphaWallet/Localization/en.lproj/Localizable.strings
Comment (James-Sangalli) [inline AlphaWallet/Localization/en.lproj/Localizable.strings:520]: typo: should be Disconnect

**Review by James-Sangalli — CHANGES_REQUESTED** [review]
@vladyslav-iosdev followed this flow and got back an error:

1. go to app.uniswap.org
2. Connect wallet via wallet connect (connection is successful)
3. Initiate a swap request
4. Error: Wallet connect failure with payload `<== {"id":1607305048318588,"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"data":"0x7ff36ab500000000000000000000000000000000000000000000000000000000038303c7000000000000000000000000000000000000000000000000000000000000008000000000000000000000000042ea529282dde0aa87b42d9e83316eb23fe62c3f000000000000000000000000000000000000000000000000000000005fcd8bec0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7","from":"0x42ea529282dde0aa87b42d9e83316eb23fe62c3f","gas":"0x25d1b","to":"0x7a250d5630b4cf539739df2c5dacb4c659f2488d","value":"0x16345785d8a0000"}]}`

**Comment by vladyslav-iosdev** [comment]
> 4\. {"id":1607305048318588,"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"data":"0x7ff36ab500000000000000000000000000000000000000000000000000000000038303c7000000000000000000000000000000000000000000000000000000000000008000000000000000000000000042ea529282dde0aa87b42d9e83316eb23fe62c3f000000000000000000000000000000000000000000000000000000005fcd8bec0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7","from":"0x42ea529282dde0aa87b42d9e83316eb23fe62c3f","gas":"0x25d1b","to":"0x7a250d5630b4cf539739df2c5dacb4c659f2488d","value":"0x16345785d8a0000"}]}

it seems like there is no `gasPrice` value in json, i will update code to handle that.

**Comment by James-Sangalli** [comment]
@vladyslav-iosdev this also fails (it is an approve transaction to uniswap):

`WC: <== {"id":1607399607708934,"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"data":"0x095ea7b30000000000000000000000007a250d5630b4cf539739df2c5dacb4c659f2488dffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff","from":"0x42ea529282dde0aa87b42d9e83316eb23fe62c3f","gas":"0xc6ee","to":"0xdac17f958d2ee523a2206206994597c13d831ec7"}]}`

**Comment by James-Sangalli** [comment]
> > 4. {"id":1607305048318588,"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"data":"0x7ff36ab500000000000000000000000000000000000000000000000000000000038303c7000000000000000000000000000000000000000000000000000000000000008000000000000000000000000042ea529282dde0aa87b42d9e83316eb23fe62c3f000000000000000000000000000000000000000000000000000000005fcd8bec0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7","from":"0x42ea529282dde0aa87b42d9e83316eb23fe62c3f","gas":"0x25d1b","to":"0x7a250d5630b4cf539739df2c5dacb4c659f2488d","value":"0x16345785d8a0000"}]}
> 
> it seems like there is no `gasPrice` value in json, i will update code to handle that.

tried a swap again and it still says wallet connect failure (however the first time it worked with a confirmation popup, second time it did not): 

`WC: <== {"id":1607399798753773,"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"data":"0x38ed17390000000000000000000000000000000000000000000000000de0b6b3a764000000000000000000000000000000000000000000000000000000000000000f872600000000000000000000000000000000000000000000000000000000000000a000000000000000000000000042ea529282dde0aa87b42d9e83316eb23fe62c3f000000000000000000000000000000000000000000000000000000005fcefdfc00000000000000000000000000000000000000000000000000000000000000030000000000000000000000006b175474e89094c44da98b954eedeac495271d0f000000000000000000000000c00e94cb662c3520282e6f5717214004a7f26888000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48","from":"0x42ea529282dde0aa87b42d9e83316eb23fe62c3f","gas":"0x36b83","to":"0x7a250d5630b4cf539739df2c5dacb4c659f2488d"}]}`

**Comment by vladyslav-iosdev** [comment]
@James-Sangalli its strange, it seems like u trying to send transaction without `value` param.

**Comment by James-Sangalli** [comment]
> @James-Sangalli its strange, it seems like u trying to send transaction without `value` param.

Yes, it has no eth value but we should be able to tolerate missing fields like that by setting them to a default of zero for example

**Comment by vladyslav-iosdev** [comment]
@James-Sangalli  please check latest changes, now it should work

**Comment by James-Sangalli** [comment]
> @James-Sangalli please check latest changes, now it should work

@vladyslav-iosdev better but still failing with the approve payload.

We can tolerate the following fields being missing:
- value (default to zero)
- gas (can get estimates from the node if unavailable)
- gas limit - as above we can get that with functions we already have 
- gas price - as above

**Comment by James-Sangalli** [comment]
@vladyslav-iosdev please also fix build issue with failing test on travis

**Comment by vladyslav-iosdev** [comment]
> @vladyslav-iosdev please also fix build issue with failing test on travis

now it should work

**Comment by vladyslav-iosdev** [comment]
> > @James-Sangalli please check latest changes, now it should work
> 
> @vladyslav-iosdev better but still failing with the approve payload.
> 
> We can tolerate the following fields being missing:
> 
> * value (default to zero)
> * gas (can get estimates from the node if unavailable)
> * gas limit - as above we can get that with functions we already have
> * gas price - as above

its strange but, all these fields non required and can be nil, can u give me detailed information, where it fails
as i understand  failure happens only for sendTransaction  action?

**Comment by James-Sangalli** [comment]
> > > @James-Sangalli please check latest changes, now it should work
> > 
> > 
> > @vladyslav-iosdev better but still failing with the approve payload.
> > We can tolerate the following fields being missing:
> > 
> > * value (default to zero)
> > * gas (can get estimates from the node if unavailable)
> > * gas limit - as above we can get that with functions we already have
> > * gas price - as above
> 
> its strange but, all these fields non required and can be nil, can u give me detailed information, where it fails
> as i understand failure happens only for sendTransaction action?

@vladyslav-iosdev the transactions are going through but they are making a fatal error, they are creating contracts instead of calling functions at an existing address, for example: https://etherscan.io/tx/0x4bdcba7e42796592b8b794363cbf5b06095cee70c932a346c15f1120d3efcba9 (this was done via wallet connect on uniswap)

You need to make sure the transaction to field is filled in properly else it will do this

**Comment by vladyslav-iosdev** [comment]
@James-Sangalli can u send me some eth to this addres `0x5d4Fdf90c45a7310e7DeAB4471038C067c2BdfCE` to test this case ?

File: AlphaWallet/WalletConnect/Types/UnconfirmedTransaction+Extensions.swift
Comment (James-Sangalli) [inline AlphaWallet/WalletConnect/Types/UnconfirmedTransaction+Extensions.swift:21]: @vladyslav-iosdev this was causing the issue whereby the to value is not set and it therefore makes a new contract

**Comment by James-Sangalli** [comment]
> @James-Sangalli can u send me some eth to this addres `0x5d4Fdf90c45a7310e7DeAB4471038C067c2BdfCE` to test this case ?

Handled with commit 7327a54

**Comment by James-Sangalli** [comment]
@vladyslav-iosdev @hboon rebased with master to fix conflict, have one more go with it before merging

File: AlphaWallet/WalletConnect/Types/UnconfirmedTransaction+Extensions.swift
Comment (James-Sangalli) [reply AlphaWallet/WalletConnect/Types/UnconfirmedTransaction+Extensions.swift:21]: @vladyslav-iosdev could we not just remove the contract field and simply use the recipient? I see no need for the difference

**Comment by James-Sangalli** [comment]
@vladyslav-iosdev please fix the build errors

**Review by James-Sangalli — APPROVED** [review]
@vladyslav-iosdev please rebase first, @hboon all good from your side?

**Comment by hboon** [comment]
@vladyslav-iosdev add changing the commit message to "Add WalletConnect support" would be more appropriate now.

**Comment by hboon** [comment]
Some observations, but let's not fix/add them until after the PR is merged, let's get this merged:

* I tested with Uniswap, but when I sent an approve transaction, it doesn't show a pending transaction in Activity. The transaction is processed successfully though.
* Maybe we should create a new TransactionConfirmationConfiguration enum-case instead of reuse the one for dapp. I haven't read through the code even to determine this though
* Seems to persist or at least try to across app restarts. But will have to test this a bit more
* Value of gas not included in actionsheet
* Not decoding approvals and displaying them as such in the actionsheet (but this is because the codebase doesn't support it yet, not because of the WalletConnect functionality in this PR)
* Should probably change where the "WalletConnect" banner/row appears
* Support multiple sessions at a time
* Can't tell if testnet in the WalletConnect session banner and screen. URL doesn't say too. Not sure if it's something we can do
* It seems to connect using the current network settings? The concept of "current network" is only used by the dapp browser, since we support multiple networks at once, will have to check if this can be improve. The problem is if we have mainnet disabled and use WalletConnect, the dapp will complain "Wrong network" and then even after enabling mainnet, it wouldn't work even if we go to the dapp browser and switch to mainnet. Might be partly uniswap though

Will create follow up issues.

**Comment by hboon** [comment]
> @vladyslav-iosdev please rebase first, @hboon all good from your side?

@James-Sangalli Looks good otherwise. Let's merge after rebase.

**Review by hboon — APPROVED** [review]

**Comment by hboon** [comment]
@James-Sangalli Rebased and force pushed

**Pull request merged**
